// hapkido-server/lib/static.js
const express = require('express');
const rollup  = require('express-middleware-rollup');
const sass    = require('node-sass-middleware');
const buble   = require('rollup-plugin-buble');

const path    = require('path');
const pkgRoot = path.normalize(path.join(__dirname, '../'));

const composeMiddlewares = (...middlewares) => middlewares.reduce(
  (a, b) =>
    (req, res, next) =>
      a(req, res, err => err ? next(err) : b(req, res, next))
);

// Javascript compilation
const js = rollup({
  src: 'frontend/js',
  dest: 'static/js',
  root: pkgRoot,
  prefix: '/js',
  rollupOpts: {
    plugins: [buble()]
  }
});

// Css compilation
const scss = sass({
  root: pkgRoot,
  src: 'frontend/css',
  dest: 'static/css',
  outputStyle: 'extended',
  prefix: '/css',
  // TODO maybe we want to use https://www.npmjs.com/package/find-node-modules here
  includePaths: [
    `${pkgRoot}/..`
  ]
});

// Static file server
const statics = express.static(`${pkgRoot}/static`);

module.exports = composeMiddlewares(js, scss, statics);
